src/geometry/manifold/times_cont_mdiff_map.lean

```diff
@@ -0,0 +1,92 @@
+/-
+Copyright © 2020 Nicolò Cavalleri. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author: Nicolò Cavalleri.
+-/
+
+import geometry.manifold.times_cont_mdiff
+import topology.continuous_map
+
+/-!
+# Smooth bundled map
+
+In this file we define the type `times_cont_mdiff_map` of `n` times continuously differentiable
+bundled maps.
+-/
+
+variables {𝕜 : Type*} [nondiscrete_normed_field 𝕜]
+{E : Type*} [normed_group E] [normed_space 𝕜 E]
+{E' : Type*} [normed_group E'] [normed_space 𝕜 E']
+{H : Type*} [topological_space H]
+{H' : Type*} [topological_space H']
+(I : model_with_corners 𝕜 E H) (I' : model_with_corners 𝕜 E' H')
+(M : Type*) [topological_space M] [charted_space H M] [smooth_manifold_with_corners I M]
+(M' : Type*) [topological_space M'] [charted_space H' M'] [smooth_manifold_with_corners I' M']
+{E'' : Type*} [normed_group E''] [normed_space 𝕜 E'']
+{H'' : Type*} [topological_space H'']
+{I'' : model_with_corners 𝕜 E'' H''}
+{M'' : Type*} [topological_space M''] [charted_space H'' M''] [smooth_manifold_with_corners I'' M'']
+(n : with_top ℕ)
+
+/-- Bundled `n` times continuously differentiable maps. -/
+@[protect_proj]
+structure times_cont_mdiff_map :=
+(to_fun                   : M → M')
+(times_cont_mdiff_to_fun  : times_cont_mdiff I I' n to_fun)
+
+/-- Bundled smooth maps. -/
+@[reducible] def smooth_map := times_cont_mdiff_map I I' M M' ⊤
+
+notation `C[` n `](` I `, ` M `; ` I' `, ` M' `)` := times_cont_mdiff_map I I' M M' n
```

Comment:
Lean doesn't allow/like an opening parenthesis `(` as a token in notation.

This notation is bad though: after declaring this notation, whenever Lean sees `](` it will treat them as a single token, which is problematic if you meant to write `] (`. So for example, this example will fail after this notation is declared, because there is no space between `]` `(`.
```
example {α} [group α](x : α) : x = x := rfl
```

Reply:
Ok do you have suggestions for a better notation? I am not really sure about how to invent a better one!

Reply:
If you reorder the arguments of `times_cont_mdiff_map`, you could do

```
notation `C[` n `]` := times_cont_mdiff_map n
```
and then you can write `C[n] I M I' M'`. Would that be acceptable?

Reply:
Oh, that doesn't work with the notations below this...
You could try `C^n[ I, M; I', M' ]`, That should work, but uses square brackets...

Reply:
> Oh, that doesn't work with the notations below this...
> You could try `C^n[ I, M; I', M' ]`, That should work, but uses square brackets...

If this does not create problems to Lean, it should be fine I guess (it is not consistent with the notation of continuous map but we can survive and if necessary change the continuous functions notation later on). I will change it together with Sebastien's suggestion. Maybe he'll also have an idea for a good notation.

Reply:
Maybe change to a calligraphic `C`, i.e., `𝓒` (type with \MCC) to avoid ambiguities. I think `𝓒^n[ I, M; I', M']` is fine, can you change to it?

Reply:
It's no problem to me but I would like things to be consistent with the notation for continuous functions, so if @semorrison, who I believe created the other notation, is good with this, I will change both.

Reply:
No, I think it predated me! No objection, however.

Reply:
> Maybe change to a calligraphic `C`, i.e., `𝓒` (type with \MCC) to avoid ambiguities. I think `𝓒^n[ I, M; I', M']` is fine, can you change to it?

One disadvantage of this however is that it is not anymore coherent with the standard latex notation $C^{\infty}$. It's not a big deal, but just to know, what other use does or could the symbol `C^` have?

Reply:
ok, let's go for `C^`, and we'll change it later if needed.

Reply:
> Oh, that doesn't work with the notations below this...
> You could try `C^n[ I, M; I', M' ]`, That should work, but uses square brackets...

@fpvandoorn this notation is not accepted by Lean the same way it is not accepted with any other bracket. Combinations of brackets are accepted. Which combination of brackets does not create the problems you pointed out? [( )]? {( )}? Suggestions are welcomed

Reply:
The only accepted bracket is `⦄`, extremely slow to write but acceptable. Is it fine?

Reply:
Let's go with this. For readability, you should insert spaces when you use the notation, i.e., write `C^n ⦃I, M; I', M'⦄` instead of `C^n⦃I, M; I', M'⦄`. Also, the notation for `C∞` really doesn't seem necessary to me (and there is a high risk to confuse the reader). Instead, just write `C^∞ ⦃I, M; I', M'⦄` (when `open_locale manifold` is active, `∞` means top). Finally, instead of having globally active notations, can you localize them to manifolds?

Reply:
Ok for `open_locale` but as for the space this is not too much in line with the notation for continuous functions (that I guess I should adapt to this one, especially if I remove `C∞` with round parentheses) and it forces to put parentheses around the whole thing when used as an argument, like `(C^n ⦃I, M; I', M'⦄)` which to me seems to make readability worse than just `C^n⦃I, M; I', M'⦄`.

Reply:
I mean the overall effect is
```
instance has_scalar' : has_scalar C^n⦃I, N; 𝕜⦄ C^n⦃I, N; J, V⦄
```
vs
```
instance has_scalar' : has_scalar (C^n ⦃I, N; 𝕜⦄) (C^n ⦃I, N; J, V⦄)
```

Reply:
It shouldn't be:
```
instance has_scalar' : has_scalar C^n ⦃I, N; 𝕜⦄ C^n ⦃I, N; J, V⦄
```
should work just fine (and if it doesn't it just means you need to adjust binding priorities)

Reply:
But in fact I prefer the version with parentheses, it looks more readable to me.

Reply:
I changed it but I still think it was better before for 3 reasons:
1. like this it seems C^n and ⦃I, N; 𝕜⦄ are separate arguments and to my eye it cofuses things: looking at statements before it was easy to get with a glimpse which things are together and which are not (in `has_scalar C^n ⦃I, N; 𝕜⦄ C^n ⦃I, N; J, V⦄` it seems `has_scalar` is taking 4 arguments; also things like `inhabited C^n ⦃I, M; I', M'⦄` or `has_coe (E →L[𝕜] E') C^n ⦃𝓘(𝕜, E), E; 𝓘(𝕜, E'), E'⦄` now look confusing). I really don't see any reason why the space improves readability
2. again this is not in line with the notation of continuous functions
3. this is not in line with the notation in standard maths where there is no space

Reply:
Coherence is a very important goal. We have spaces everywhere before parentheses in mathlib, so why would you want to remove them here? You even wrote yourself `has_coe (E →L[𝕜] E')` and not `has_coe(E →L[𝕜] E')`... In the same way, you should probably write `𝓘 (𝕜, E)` instead of `𝓘(𝕜, E)`, by the way -- or even better `𝓘 𝕜 E`, again for coherence. The goal is not to mimick the usual mathematical notation at all cost.

Reply:
Indeed I want to remove them because of coherence!! The convention in mathlib seems to be that notation involving just one character and parentheses that are part of the notation do not have spaces in between them (I think this is very reasonable but it is not me who chose this, it has been like this since ever. Just to cite some examples:
- `C(X, Y)` denotes continuous functions and does not have a space (and coherence with this is very important here.
- `𝓝[s]` denotes neighborhood within and is written without space
- `X⟦n⟧` is notation for shift in category theory, written without space
- `f⟦n⟧` as above does not have a space
- morphisms like `→L[𝕜]` involving one letter and a parenthesis do not have a space (they are not `→L [𝕜]`)
- `![a, b, ...]` does not have a space
- A very similar case to our `ℤ_[p]` or `ℚ_[p]` do not have spaces
I can list you other examples but you got the idea!

I think that if we decide not to follow this convention here we should propose to change it everywhere else in mathlib as well.